### DISCUSSION TITLE
Different remark AST compared to built in plugins

### DISCUSSION THREAD

**Discussion by anatolykopyl** [facebook/docusaurus#9679] [Q&A] [answered]
I use the source code of Docusaurus as an example, when trying to create my own remark plugin and although it seems that my plugin is used the same way in [processor.ts](https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-mdx-loader/src/processor.ts), it can not visit some nodes the same way the default plugins can.

For example where the [head plugin](https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-mdx-loader/src/remark/head/index.ts) visits `mdxJsxFlowElement`, my plugin does not visit a single element. Instead I can visit nodes of type `jsx`, but they have a different structure. Same goes for `mdxjsEsm` nodes, which are replaced with nodes of type `import`.

For some reason the first argument in the returned function from my plugin gets a different tree, compared to the default plugins. This is the same for both `remarkPlugins` and `beforeDefaultRemarkPlugins` options.

Thank you in advance!

**Comment by Josh-Cena** [comment]
@anatolykopyl, are you using Docusaurus v3? Seems you are handling MDX v1 AST.

**Reply by anatolykopyl** [reply]
Thank you for the fast response! 
No, I am stuck on Docusaurus v2 for now. The difference is because of the MDX version? Is there a way to get a tree similar to the one I described without upgrading to Docusaurus v3?

**Reply by Josh-Cena** [reply, marked as answer]
No; MDX v1 uses a different tree. Read examples here instead: https://github.com/facebook/docusaurus/tree/docusaurus-v2/packages/docusaurus-mdx-loader/src Curiously, https://astexplorer.net/ still uses the MDX v1 parser, so you can test there.

**Comment by johnnyreilly** [comment]
Not quite what you're after, but this blog post (and the related ones) will give you a sense of the different AST that old and new MDX have. I say it's not quite what you're after as these are rehype plugins, but I thought I'd share on the off chance 

https://johnnyreilly.com/docusaurus-3-how-to-migrate-rehype-plugins
